pkg/karmadactl/promote/promote.go

```diff
@@ -249,12 +265,121 @@ func (o *CommandPromoteOption) Run(f util.Factory, args []string) error {
 		return fmt.Errorf("failed to get gvr from %q: %v", o.gvk, err)
 	}
 
-	return o.promote(controlPlaneRestConfig, obj, gvr)
+	if o.Deps {
+		err := o.promoteDeps(memberClusterFactory, obj, mapper, gvr, controlPlaneRestConfig)
+		if err != nil {
+			return fmt.Errorf("failed to promote dependencies of resource %q(%s/%s) automatically: %v", gvr, o.Namespace, o.name, err)
+		}
+	}
+
+	return o.promote(controlPlaneRestConfig, obj, gvr, false)
+}
+
+// doPromoteDeps promotes dependencies of the resource
+func (o *CommandPromoteOption) doPromoteDeps(memberClusterFactory cmdutil.Factory, dependencies []configv1alpha1.DependentObjectReference, mapper meta.RESTMapper, config *rest.Config) error {
+	for _, dep := range dependencies {
+		depInfo, err := o.getObjInfo(memberClusterFactory, o.Cluster, []string{dep.Kind, dep.Name})
+		if err != nil {
+			return fmt.Errorf("failed to get dependency %q(%s/%s) in cluster(%s): %v", dep.Kind, dep.Namespace, dep.Name, o.Cluster, err)
+		}
+
+		depObj := depInfo.Info.Object.(*unstructured.Unstructured)
+		depGvk := depObj.GetObjectKind().GroupVersionKind()
+		depGvr, err := restmapper.GetGroupVersionResource(mapper, depGvk)
+		if err != nil {
+			return fmt.Errorf("failed to get dependency gvr from %q: %v", depGvk, err)
+		}
+
+		err = o.promote(config, depObj, depGvr, true)
+		if err != nil {
+			return fmt.Errorf("failed to promote dependency %q(%s/%s) in cluster(%s): %v", dep.Kind, dep.Namespace, dep.Name, o.Cluster, err)
```

Comment:
I think we need to delete the dependencies created before if one failed in the middle.
For example:
nginx-deployment depends on A/B/C configmap.
1.Promote A.
2.Promote B failed.

Then we should delete A in karmada control plane. But here is a problem, delete A will case configmap/A deletion in the member cluster.

I think we can ublabel the configmap/A to implement it.

Reply:
> I think we need to delete the dependencies created before if one failed in the middle.

🤔，I think that only some warning messages are needed. There is no need to delete the already `promote ` successful dependent resources.

Reply:
Ask for @RainbowMango 's suggestion.

Reply:
> Then we should delete A in karmada control plane. But here is a problem, delete A will case configmap/A deletion in the member cluster.

This is concerning, it depends on if we have created the `PP`, if not, then cascade deletion would not happen.

> 🤔，I think that only some warning messages are needed. There is no need to delete the already promote  successful dependent resources.

Both are OK to me, but if not clean the resources, we should ensure no errors during the next promotion. (Ignore already existing errors).

I tend to make a cleanup, that is revert all changes made by this command.

Reply:
> I tend to make a cleanup, that is revert all changes made by this command.

+1